File: Semester 6/MA 374 (Financial Engg. Lab)/Lab 3/180123062_AB_q3_2.cpp
// Question 3 Algorithm 2 Lab Assignment 03
// @AB Satyaprakash, 180123062

// include all libraries
#include <bits/stdc++.h>
using namespace std;

map<pair<double, double>, double> mp;

// functions
double recursion(double n, double K, double S, double T, double r, double sig, double M, double u, double d, double p, double q, double dt, int upCnt)
{
    // if already in the map, then return
    if (mp.count({n, upCnt}))
        return mp[{n, upCnt}];

    // else if n==M and assign to map and return
    if (n == M)
    {
        mp[{n, upCnt}] = max(S - K, 0.0);
        return max(S - K, 0.0);
    }

    // else do recursion
    double up = recursion(n + 1, K, S * u, T, r, sig, M, u, d, p, q, dt, upCnt + 1);
    double dn = recursion(n + 1, K, S * d, T, r, sig, M, u, d, p, q, dt, upCnt);
    double pc = (exp(-r * dt)) * (p * up + q * dn);
    mp[{n, upCnt}] = pc;
    return pc;
}

// code begins in main
int main()
{
    // Given initial values and taking K = 100
    double S0 = 100, K = 100, T = 1, r = 0.08, sig = 0.2;

    vector<double> M(4);
    // Can handle M values upto 1000
    M = {5, 10, 25, 50, 100, 500, 1000};

    for (int i = 0; i < (int)M.size(); i++)
    {
        // evaluate dt, u, d, p and q as per formulae given in question
        double dt = T / M[i];
        double u = exp(sig * sqrt(dt) + (r - sig * sig / 2) * dt);
        double d = exp(-sig * sqrt(dt) + (r - sig * sig / 2) * dt);
        double p = (exp(r * dt) - d) / (u - d);
        double q = 1 - p;
        mp.clear();

        double optionPrice = recursion(0, K, S0, T, r, sig, M[i], u, d, p, q, dt, 0);
        cout << fixed << setprecision(15);
        cout << "The initial european call option price for M = " << (int)M[i] << " is " << optionPrice << endl;
    }

    return 0;
}

File: Semester 6/MA 374 (Financial Engg. Lab)/Lab 3/180123062_AB_q3_2.py
# Question 3 Algorithm 2 (efficient) Lab Assignment 03
# @AB Satyaprakash, 180123062

# imports
from math import exp, sqrt
import numpy as np
import time

# Initialising a dictionary -- C++ equivalent to map<>
Map = {}
# functions ---------------------------------------------------------------------------


def recursion(n, K, S, T, r, sig, M, u, d, p, q, dt, upCnt):
    if (n, upCnt) in Map:
        return Map[(n, upCnt)]

    if n == M:
        Map[(n, upCnt)] = max(S - K, 0)
        return max(S - K, 0)

    up = recursion(n + 1, K, S * u, T, r, sig, M, u, d, p, q, dt, upCnt+1)
    dn = recursion(n + 1, K, S * d, T, r, sig, M, u, d, p, q, dt, upCnt)
    pc = (exp(-r * dt)) * (p * up + q * dn)
    Map[(n, upCnt)] = pc
    return pc


# -------------------------------------------------------------------------------------
# Given Initial Values -- and taking K = 100
S0, K, T, r, sig = 100, 100, 1, 0.08, 0.2
# Can handle values of M as much as 500 ..
Mlist = [5, 10, 25, 50, 100, 500]

for M in Mlist:
    dt = T/M

    u = exp(sig*sqrt(dt)+(r-sig*sig/2)*dt)
    d = exp(-sig*sqrt(dt)+(r-sig*sig/2)*dt)
    p = (exp(r*dt)-d)/(u-d)
    q = 1-p
    Map.clear()
    loopbackOptionPrice = recursion(0, S0, S0, T, r, sig, M, u, d, p, q, dt, 0)
    print('The initial european call option price for M = {} is {}'.format(M, loopbackOptionPrice))

# Test computation time for M = 15: -----------------------------------------------------
dt = T/M
u = exp(sig*sqrt(dt)+(r-sig*sig/2)*dt)
d = exp(-sig*sqrt(dt)+(r-sig*sig/2)*dt)
p = (exp(r*dt)-d)/(u-d)
q = 1-p
Map.clear()
start = time.time()
timeTempPrice = recursion(0, S0, S0, T, r, sig, M, u, d, p, q, dt, 0)
end = time.time()

print('Computational time for M = 15 is {}s'.format(end-start))

# Question 3 algorithm 2 ends -------------------------------------------------------------
